GTG/core/tags2.py

```diff
@@ -0,0 +1,208 @@
+        """String representation."""
+
+        return f'Tag Store. Holds {len(self.lookup)} tag(s)'
+
+
+    def find(self, name: str) -> Tag2:
+        """Get a tag by name."""
+
+        return self.lookup_names[name]
+
+
+    def new(self, name: str, parent: UUID = None) -> Tag2:
+        """Create a new tag and add it to the store."""
+
+        name = name if not name.startswith('@') else name[1:]
+
+        try:
+            return self.lookup_names[name]
+        except KeyError:
+            tid = uuid4()
+            tag = Tag2(id=tid, name=name)
+
+            if parent:
+                self.add(tag, parent)
+            else:
+                self.data.append(tag)
+                self.lookup[tid] = tag
+                self.lookup_names[name] = tag
+
+            self.emit('added', tag)
+            return tag
+
+
+    def from_xml(self, xml: Element) -> None:
+        """Load searches from an LXML element."""
+
+        elements = list(xml.iter(self.XML_TAG))
+
+        # Do parent searches first
+        for element in elements.copy():
+
+            tid = element.get('id')
+            name = element.get('name')
+            color = element.get('color')
+            icon = element.get('icon')
+
+            tag = Tag2(id=tid, name=name)
+            tag.color = color
+            tag.icon = icon
+
+            self.add(tag)
+
+            log.debug('Added %s', tag)
+
+
+        for element in elements:
+            parent_name = element.get('parent')
+
+            if parent_name:
+                tid = element.get('id')
+
+                try:
+                    parent = self.find(parent_name)
+                    self.parent(tid, parent.id)
+                    log.debug('Added %s as child of %s', tag, parent)
+                except KeyError:
+                    pass
+
+
+    def to_xml(self) -> Element:
+        """Save searches to an LXML element."""
+
+        root = Element('SavedSearches')
+
+        parent_map = {}
+
+        for tag in self.data:
+            for child in tag.children:
+                parent_map[child.id] = tag.id
+
+        for tag in self.lookup.values():
+            element = SubElement(root, self.XML_TAG)
+            element.set('id', str(tag.id))
+            element.set('name', tag.name)
+
+            if tag.color:
+                element.set('color', tag.color)
+
+            if tag.icon:
+                element.set('icon', tag.icon)
+
+            try:
+                element.set('parent', str(parent_map[tag.id]))
+            except KeyError:
+                pass
+
+        return root
+
+
+    def generate_color(self) -> Color:
+        """Generate a random color that isn't already used."""
+
+        MAX_VALUE = 65535
+        color = None
+
+        while color in self.used_colors:
+            color = Color(
+                random.randint(0, MAX_VALUE),
+                random.randint(0, MAX_VALUE),
+                random.randint(0, MAX_VALUE)
+            ).to_string()
```

Comment:
hum, this changes the string representation since `to_string()` in `RGBA` returns something like `rgb(x,x,x)`. We would need to break compat in the xml format.
I think we should do this later, once we also have the new recurring task code and we definitely break compat (so we can load() current gtg_data.xml to test things)

Reply:
This is how I generate the colour string (could maybe be moved to a util file if we have one): https://github.com/getting-things-gnome/gtg/blob/accd2d88b80f8b6f7fa36cc175d97b9c20d64ec4/GTG/gtk/browser/tag_editor.py#L229-L231